library/alloc/src/rc.rs

```diff
@@ -1002,7 +1002,9 @@ impl<T: ?Sized> Rc<T> {
     ///
     /// [`ptr::eq`]: core::ptr::eq
     pub fn ptr_eq(this: &Self, other: &Self) -> bool {
-        this.ptr.as_ptr() == other.ptr.as_ptr()
+        // The *const u8 cast discards the vtable to work around
+        // https://github.com/rust-lang/rust/issues/46139
```

Comment:
This is misleading, as this is not just because of that issue. If we're deciding now that `ptr_eq` compares the allocation (like the documentation says), we will not change this back if that issue ever gets solved. This change will make the example in https://github.com/rust-lang/rust/pull/80505#issuecomment-752977080 return `true` on purpose. (Which would also make a good test case.)

Reply:
My understanding was that we’re deeming that example an incorrect use of `transmute` (as per https://github.com/rust-lang/rust/pull/80505#issuecomment-757213210), and thus making no guarantee whether it returns `true` or `false`?

Reply:
@m-ou-se the author needs help, after that we can move this ahead

Reply:
@m-ou-se As I understand it, your statement is incorrect. There are no other reasons besides https://github.com/rust-lang/rust/issues/46139 for doing this.  

In fact, after a better solution to https://github.com/rust-lang/rust/issues/46139 appears then reverting this change will reduce bugs.